### ISSUE TITLE
On custom data, loss and weights diverge to nan -- faster with lower LR

### ISSUE THREAD

**Issue by anandthakker** [alexgkendall/caffe-segnet#31]
Thanks very much for sharing your work on SegNet!

I'm interested in using the [segnet model](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Models/segnet_train.prototxt) with some custom data (aerial imagery; 1000 images and 3 label classes for now), and I'm seeing something strange in training:

Training with `base_lr: 0.001`, I was seeing training loss = `nan`.  Surprisingly, when I _reduced_ the LR, the loss exploded to `nan` even _faster_.  Couple of examples:

With base_lr = 1e-4

```
#Iters Seconds TrainingLoss LearningRate
0    3.022133     1.25783   0.0001
20   58.911155    1.3897    0.0001
40   114.768299   1.18601   0.0001
60   170.620245   1.02438   0.0001
80   226.502911   0.992765  0.0001
100  282.332283   1.05062   0.0001
120  338.179408   1.44935   0.0001
140  394.035362   4.73935   0.0001
160  449.878457   0.997907  0.0001
180  505.701507   0.98053   0.0001
200  561.558644   1.14603   0.0001
220  617.398724   0.936918  0.0001
240  672.909704   nan       0.0001
260  728.026581   nan       0.0001
```

With base_lr = 1e-6

```
#Iters Seconds TrainingLoss LearningRate
0   3.033944    1.4484   1e-06
20  58.966689   1.12416  1e-06
40  114.393818  nan      1e-06
60  169.507581  nan      1e-06
80  224.611446  nan      1e-06
```

I also set `debug_info: true`, and discovered that the actual _weights_ seem to be diverging, also:

```
I0408 13:13:13.229012 23306 net.cpp:594]     [Forward] Layer data, top blob data data: 62.8498
I0408 13:13:13.229233 23306 net.cpp:594]     [Forward] Layer data, top blob label data: 2.95128
I0408 13:13:13.229307 23306 net.cpp:594]     [Forward] Layer label_data_1_split, top blob label_data_1_split_0 data: 2.95128
I0408 13:13:13.229379 23306 net.cpp:594]     [Forward] Layer label_data_1_split, top blob label_data_1_split_1 data: 2.95128
I0408 13:13:13.236066 23306 net.cpp:594]     [Forward] Layer conv1_1, top blob conv1_1 data: nan
I0408 13:13:13.236177 23306 net.cpp:604]     [Forward] Layer conv1_1, param blob 0 data: nan
I0408 13:13:13.236227 23306 net.cpp:604]     [Forward] Layer conv1_1, param blob 1 data: nan
I0408 13:13:13.258630 23306 net.cpp:594]     [Forward] Layer conv1_1_bn, top blob conv1_1 data: nan
I0408 13:13:13.258780 23306 net.cpp:604]     [Forward] Layer conv1_1_bn, param blob 0 data: nan
I0408 13:13:13.258839 23306 net.cpp:604]     [Forward] Layer conv1_1_bn, param blob 1 data: nan
I0408 13:13:13.261641 23306 net.cpp:594]     [Forward] Layer relu1_1, top blob conv1_1 data: 0
I0408 13:13:13.304590 23306 net.cpp:594]     [Forward] Layer conv1_2, top blob conv1_2 data: nan
...
```

Finally, if I set `base_lr` to 0, the loss diverges, but the weights do not.

I'm going to keep investigating, and will report back here if I resolve it, but I'm a bit stumped.  @alexgkendall Do any insights come to mind?

**Comment by alexgkendall** [comment]
Hey - some questions:
- Do you use any class balancing (or how unbalanced is your dataset)?
- Do you observe this on another dataset, eg. CamVid?
- What happens to the individual class accuracies before loss goes to inf?

**Comment by anandthakker** [comment]
> Do you observe this on another dataset, eg. CamVid?

No, although I've only run it for a couple hundred iterations.  I'll go back and try running CamVid to convergence.

> Do you use any class balancing (or how unbalanced is your dataset)?
> What happens to the individual class accuracies before loss goes to inf?

I am using class balancing, _but_ your asking did help me notice a bug in my data prep scripts which had led to the per-class weights being wrong.  Specifically, the third class was weighted too high (~8 instead of ~2).  Probably this relates to your other question: in both of the logs from my original question, class accuracies go to 0,0,1 (for class numbers 0, 1, 2 resp.) in the same iteration where loss goes to inf.

I've just done a new run with corrected class weights, and the same thing ended up happening, although it did take about twice as long:

```
I0411 10:35:22.930238  2418 solver.cpp:214] Iteration 500, loss = 1.28299
I0411 10:35:22.930279  2418 solver.cpp:229]     Train net output #1: loss = 1.28299 (* 1 = 1.28299 loss)
I0411 10:35:22.930294  2418 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.355205
I0411 10:35:22.930308  2418 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.490196
I0411 10:35:22.930320  2418 solver.cpp:229]     Train net output #4: per_class_accuracy = -nan

I0411 10:36:18.806130  2418 solver.cpp:214] Iteration 520, loss = 0.773778
I0411 10:36:18.806154  2418 solver.cpp:229]     Train net output #1: loss = 0.773779 (* 1 = 0.773779 loss)
I0411 10:36:18.806161  2418 solver.cpp:229]     Train net output #2: per_class_accuracy = 0.657882
I0411 10:36:18.806167  2418 solver.cpp:229]     Train net output #3: per_class_accuracy = 0.451116
I0411 10:36:18.806174  2418 solver.cpp:229]     Train net output #4: per_class_accuracy = -nan

I0411 10:37:14.077673  2418 solver.cpp:214] Iteration 540, loss = nan
I0411 10:37:14.077733  2418 solver.cpp:229]     Train net output #1: loss = 80.5069 (* 1 = 80.5069 loss)
I0411 10:37:14.077757  2418 solver.cpp:229]     Train net output #2: per_class_accuracy = 0
I0411 10:37:14.077774  2418 solver.cpp:229]     Train net output #3: per_class_accuracy = 0
I0411 10:37:14.077781  2418 solver.cpp:229]     Train net output #4: per_class_accuracy = -nan

I0411 10:38:09.197612  2418 solver.cpp:214] Iteration 560, loss = nan
I0411 10:38:09.197638  2418 solver.cpp:229]     Train net output #1: loss = 81.6976 (* 1 = 81.6976 loss)
I0411 10:38:09.197645  2418 solver.cpp:229]     Train net output #2: per_class_accuracy = 0
I0411 10:38:09.197657  2418 solver.cpp:229]     Train net output #3: per_class_accuracy = 0
I0411 10:38:09.197664  2418 solver.cpp:229]     Train net output #4: per_class_accuracy = 1
```

@alexgkendall I've also been trying segnet_basic on the same dataset, and have _not_ seen this problem, although I'm not having much luck with training either:

![combined](https://cloud.githubusercontent.com/assets/4974151/14424981/34c78d00-ffb2-11e5-89dc-b918b8183cf0.png)

(in this plot 1 epoch = 50 iterations = 800 images)

**Comment by alexgkendall** [comment]
If you haven't balanced the classes effectively then the network may go unstable. Sometimes I've observed the first sign of this is the network converging to one class, followed by loss divergence.

This seems fairly extreme however, are your classes very, very unbalanced? Maybe try increasing the class loss weighting further? 

You could also try lowering the learning rate and/or using an adaptive gradient descent technique.

**Comment by anandthakker** [comment]
Thanks @alexgkendall -- I appreciate the suggestions!

I don't have too much experience with this, would you call the stats below "very, very unbalanced"?

```
class,frequency,image_count
0,985526,417
1,1272623,471
2,121344,108
3,63156507,1000
```

`frequency` is just the raw number of pixels, `image_count` is the number of images where at least one pixel of the class is present.  class 3 is 'unknown', which i'm ignoring in the loss layer.

> Maybe try increasing the class loss weighting further?

Currently, I'm calculating the losses as described in one of your papers.  (For reference: `f(class) = frequency(class) / (image_count(class) * 65536)` for each class--excluding background--and then taking `(median of f(class)) / f(class)` as the weight.)  Are you essentially suggesting reducing the weight for the class that the network is converging to?

**Comment by LukaLan** [comment]
Dear @anandthakker  and @alexgkendall 

First, thank you for your work and explanations.
I am also trying to train the model on my own data and as I am completely new in this field I have faced some problems and have some questions.

First, my set consists of only 1 class. So a class  (an object) and background.
The train and test folders consist of images (360 x 480 x 3 RGB) in .PNG format and the trainannot and testannot consist of grayscale images (360 x 480) in .PNG format. I have attached an example of an annotation image. It is a black image (0) with a value (123) where the object is present. Like this:

![9](https://cloud.githubusercontent.com/assets/9202473/15397275/cfe70a04-1ddf-11e6-8ed0-f7a6d288129c.png)
**Edit: ok, just realized that this is probably not ok. So I set the value of my object to 1 instead of 123.**

So my first question would be, is this a correct form of an annotation image?

My next question is a bit less specific. What else needs to be changed in order to successfully train your on your own dataset? I've only been trying to change the penultimate (SoftmaxWithLoss) layer. I am using the equation from the paper, the one stated above. Do I also have to put a weigh for background and in what sequence does this have to be written? Even with this done I believe there are some more changes to be done to the model. Could you please help me out?

**Comment by anandthakker** [comment]
@LukaLan I believe you should actually use `0` for the object class, and `1` for the background, and then you'll also need to set the `num_output` for the final convolution layer to `1` (instead of `11`, if you're copying the one from SegNet tutorial).  I believe you would also want to set `ignore_label: 1` (again, instead of `11`) in the loss layer.

A note: I was recently trying something similar -- that is, using SegNet with only a single label class -- and was getting loss = 0 immediately.  I don't _think_ it was a problem with my data, so I wondered whether perhaps the loss layer only works with > 1 (non-ignored) class.  @alexgkendall does this ring any bells for you?

**Comment by alexgkendall** [comment]
Hey - it doesn't make sense to train segnet on a single class. In this situation, segnet will simply learn to predict every pixel as this class, and you will see loss=0.

Instead, you need to have a background class that **isn't ignored**. It is essential to have a background class to learn what isn't your object of interest.

Therefore you should set `num_output: 2`, label your object and background classes as 0 and 1 and have no ignore label.

Cheers.

**Comment by anandthakker** [comment]
Oof, thanks @alexgkendall -- this was a pretty silly mistake on my part :-/

**Comment by LukaLan** [comment]
Dear @alexgkendall and @anandthakker, thank you very much for your input.
A quick update regarding this topic. I was able to train and test SegNet on my class, used the settings mentioned above (num_output: 2 and 0 and 1 for classes).

It all worked well but I had to tweak the weight decay in order to be able to train the first 1.000 iterations. My question is, is there a formula for calculating the optimal training parameters like learning rate, weight decay?

**Comment by anandthakker** [comment]
Noting here that I haven't seen the original issue--loss/weights diverging to nan--ever since I pulled together a new set of training data; I'm going to close this issue under the assumption that this probably had to do with bad balancing in my previous training set.  Thanks again for the advice @alexgkendall
